Autogen

Abish-27 opened the pull request:
#  Timetable Autogeneration Usecase

This PR introduces a basic auto-generation system that uses a depth-first search (DFS) backtracking algorithm to produce a valid timetable based on course offerings, locked sections, and user-defined constraints.

## Entities
- **`CourseVariable`**: One course var each containing the course offering and domain of all it's sections
- **Constraints package**: The constraints package contains constraints that need to be met that can be checked during the generation phase. Since the tutorials have not been mapped to the lectures yet I have not yet created a constraint for that but I will do that once I finish testing this, since it's quite easy to implement using this constraint system. Contains **`TimeConflictConstraint`** (avoids conflicts between 2 sections) and **`BlockedTimeConstraint`** (checks that no section's meetings collide with blocked times provided)
- **`PotentialTimetable`**: This is a class used by the interactor to conduct it's searches. It returns a possible timetable-(a timetable that has been built) and a boolean- true if all constraints are met or else false.


## Interface Adapter
- `AutogenController`: Converts UI-level selections (courses, locked sections, blocked blocks) into `AutogenInputData` and contains no business logic.
- **`AutogenPresenter`**: Bridges the gap between pure use-case output and UI state and separates success message and error message so the view can decide how to show each

## Use_Case

### Autogen Input Data
 Bundles all inputs required by the use case:
- `Set<CourseCode> selectedCourses`
- `Set<Section> lockedSections`
- `WeeklyOccupancy blockedTimes`
**Immutable DTO to keep the interactor free of UI concerns.**

### AutogenOutputData
- Holds the generated timetable
- This is what the presenter uses to update the TimetableState.

### AutogenInputBoundary / AutogenOutputBoundary
- Define the primary port interfaces for the use case and presenter
- Allow controllers and presenters to depend only on abstractions

### Autogen Interactor:
- **`execute`**: uses the AutogenInputData provided by the Controller to get necessary variables and constraint for generation algorithm to work using the AutogenDataAccessObject. Calls the `buildConstraints` and `applyLock` helpers to set up a proper constraint system.
- **`buildConstraints`**: Creates a list of `Constraints` or rules that the generated table needs to meet such as `BlockedTimeConstraint` and `TimeConflictConstraint`.
- **`applyLock`**: Goes through all the sections from the timetable that have been locked. If a certain section is locked it ensures that its CourseVariable only contains that section rather all the other sections it it's domain. This is can thought of as an automatic locking system, where it filters the locked sections for the timetable rather than introducing a new entity.
- **`generateTimetable`**: returns a `PotentialTimetable` by calling the recursive function timetableSearch
- **`timetableSearch`**: THE MVP. Uses depth first search a recursive backtracking algorithm. The ideal base case occurs when all variables have been assigned. It acts like a tree- starts with an potential timetable which is initially empty. Picks a variable, and iterates through all it's sections. For each one it tries using that section and moves ahead to the next variable, essentially forming a tree path. Each time it meets a dead end, it backtracks and tries a different section. A dead end is met after each section in a variable doesn't return a successful result. In each loop, a isConsistent( timetable) is called where we check if the  potential timetable so far is valid else we backtrack. 
- **`isConsistent`**: Takes in a Timetable and ensures that it meets all the necessary rules by looping through the list of constraints. Returns a boolean, true if all constraints are met else false. 

## UI Integration (View)
- Hooked the AutogenController into the main view and TimetablePanel
- Created an Autogenerate Button in TimetablePanel which that relies on an event listener and alerts the AutogenController when activated. 
- The Presenter's fail view (called when no timetable is generated or interactor runs into an error) creates a dialogue popup that alerts the user in the view, that a 'new' timetable could not be generated
- After successful generation the `GlobalView` (affected by `GlobalPresenter` via `AutogenPresenter`) creates a new tab on the workbook with the brand new timetable loaded in

### Data Access
- `AutogenDataAccessObject`, has a method `getSelectedCourseOfferings` which gets all the sections for a course that has been selected on the timetable. This is done by using a `JsonCourseDataRepository` object that holds the mappings to all courses and their domains. 

### Tests
- Added an AutogenInteractorTest which covers multiple scenarios such as courses having conflicts, time blocks, locks, no possible generation solution etc. (
Update: Full Coverage done in later PR- **Autogen update tests**

## Design/SOLID Notes
### Single Responsibility Principle
- AutogenController only translates user actions → use-case input.
- AutogenInteractor only does business logic / algorithm.
- AutogenPresenter only prepares view-specific state.

### Open/Closed Principle
New constraints (e.g., professor preference, campus location) can be added by:
- Implementing the Constraint interface.
- Plugging the new constraint into the constraint list used by DFS.
- The DFS engine doesn’t need to be rewritten; it just asks each constraint if a partial assignment is valid.

### Dependency Inversion Principle
- UI depends on AutogenInputBoundary rather than concrete AutogenInteractor.
- Interactor depends on AutogenOutputBoundary, not on a concrete view.
- Data sources come through data access interfaces/objects so the interactor is decoupled from any particular data storage

## TODO;
- For now you only get assigned one section per course offering, which might mean only 1 tut/lec. I will fix this once we link tutorials to lectures in the data.


### **TLDR; Finished Autogen Implementation, and applied clean architecture, need to finish data  tut-section mapping**

Abish-27 committed:
Created an autogen folder with variable and assignment classes that will be used for autgogen DRV logic.

Abish-27 committed:
Created files for Dfs and its search result.

Abish-27 committed:
Created and completed basic DfsBacktracker that makes use of assignment, search result and variable to recursively build a valid timetable.

Abish-27 committed:
Changed the constraint system.
Got rid of LockedSectionConstraint.java and replaced it with a VariableFactory that builds the required variables in such a way that if something has locked sections only those appear. This saves times and doesn't add wrong sections and go down the wrong path for no reason.

Abish-27 committed:
Created a file for the BlockedTimeConstraint class

Abish-27 committed:
Finished creating code for BlockedTimeConstraint class.

Abish-27 committed:
Got rid of redundant class Assignment.java and renamed SearchResult.java to PotentialTimetable.java as per SeriousGuy's Comments.

Abish-27 committed:
Merge branch 'blocked-times-constraint' into autogen
Adds the blocked time constraint to the main PR so we can review it as a whole

Abish-27 committed:
Refactored all autogen/constraint objects into entity folder. Began turning autogen use case into Clean architecture by implementing InputBoundary and Interactor.

Abish-27 committed:
Added a few more clean architecture files. (no code yet)

Abish-27 committed:
Finished All clean architecture classes in use case

Abish-27 committed:
DataAcessInterface for autogen

Abish-27 committed:
Finished converting Interactor to clean arch

Abish-27 committed:
Added tests for autogen interactor

Abish-27 committed:
Added data access classes and implemented input data properly for autogen

Abish-27 committed:
Finished AutogenController

Abish-27 committed:
Merge branch 'main' into autogen

SeriousGuy888 commented on src/main/java/entity/constraints/TimeConflictConstraint.java:
```diff
@@ -1,10 +1,16 @@
+    /**
+     *
+     * @param chosen the set of currently chosen sections
+     * @return true if no time intersections between all chosen sesctions else false
```
```suggestion
     * @return true if no time intersections between all chosen sections, else false
```

SeriousGuy888 commented on src/main/java/entity/CourseVariableFactory.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.util.*;
+
+public class CourseVariableFactory {
```
i'm not sure if this fits a factory pattern (e.g. from the [CA lab](CourseVariableFactory ))

though if you renamed it to something like `CourseVariableListFactory` i think it would fit?

doesn't really matter i guess, but i'd be confused to see a class named `CourseVariableFactory` only produce lists of course variables

SeriousGuy888 replied:
you could also just have this class's only method as a static method in the `CourseVariable` class

SeriousGuy888 commented on src/main/java/interface_adapter/autogen/AutogenPresenter.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void prepareSuccessView(AutogenOutputData outputData) {
+        viewModel.setTimetable(outputData.getGeneratedTimetable());   // or getSchedules()
```
what is this comment for?

SeriousGuy888 commented on src/main/java/interface_adapter/autogen/AutogenPresenter.java:
```diff
@@ -0,0 +1,27 @@
+    public void prepareSuccessView(AutogenOutputData outputData) {
+        viewModel.setTimetable(outputData.getGeneratedTimetable());   // or getSchedules()
+
+        // 2. Set a nice message for the UI to show
```
two??

```suggestion
        // Set a nice message for the UI to show
```

SeriousGuy888 commented on src/main/java/interface_adapter/autogen/AutogenController.java:
```diff
@@ -0,0 +1,22 @@
+public class AutogenController {
+    private final AutogenInputBoundary interactor;
+
+    public AutogenController(AutogenInputBoundary interactor){this.interactor=interactor;}
```
this entire constructor on one line is a bit weird and inconsistent with the rest of our code

but i think we'll be adding checkstyle later, so whoever is adding that will probably fix this if you don't change it

SeriousGuy888 commented on src/main/java/interface_adapter/autogen/AutogenPresenter.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void prepareFailView(String errorMessage) {
+        System.out.println(errorMessage);
```
left here accidentally from debugging?

```suggestion
```

i would not leave `println` in the production `AutogenPresenter`, since when other people are writing their code after this code you've written gets merged, if you have this random `println` in your presenter, they might get confused as to where this random message is printing from when they didn't add a `println` themself

SeriousGuy888 commented on src/main/java/use_case/autogen/AutogenDataAccessInterface.java:
```diff
@@ -0,0 +1,8 @@
+package use_case.autogen;
+import entity.CourseOffering;
+import  java.util.List;
```
```suggestion
import java.util.List;
```

SeriousGuy888 commented on src/main/java/use_case/autogen/AutogenInteractor.java:
```diff
@@ -0,0 +1,135 @@
+
+    /** Checks if constraints are met, using the classes in the constraints package **/
+    /**
+     *
```
invalid javadoc comment

```suggestion
    /**
     * Checks if constraints are met, using the classes in the constraints package
```

SeriousGuy888 commented on src/main/java/use_case/autogen/AutogenInteractor.java:
```diff
@@ -0,0 +1,135 @@
+    /** Checks if constraints are met, using the classes in the constraints package **/
+    /**
+     *
+     * @param assignment The timetable assinment that is being checked for consistency
```
```suggestion
     * @param assignment The timetable assignment that is being checked for consistency
```

SeriousGuy888 commented on src/main/java/use_case/autogen/AutogenOutputData.java:
```diff
@@ -0,0 +1,11 @@
+
+
+public class AutogenOutputData {
+    private final  Set<Section> generatedSections;
```
```suggestion
    private final Set<Section> generatedSections;
```

SeriousGuy888 commented on src/main/java/use_case/autogen/PotentialTimetable.java:
```diff
@@ -0,0 +1,25 @@
+
+public class PotentialTimetable {
+   boolean success;
+   Set<Section> timetable;
```
im guessing you just forgot to add access modifiers since you aren't using the package-privateness anywhere i dont think

```suggestion
    private final boolean success;
    private final Set<Section> timetable;
```

SeriousGuy888 commented on src/test/java/use_case/autogen/AutogenInteractorTest.java:
```diff
@@ -0,0 +1,364 @@
+        WeeklyOccupancy fullyBlockedTime = new WeeklyOccupancy(
+                WeeklyOccupancy.DayOfTheWeek.MONDAY,
+                1000 * 60 * 60 * 10,   // 10:00
+                1000 * 60 * 60 * 12    // 12:00
```
these are kind of magic numbers in this context

since you defined the `FakeImpossibleAutogenDataAccess` class to use these same times, but there's no way to tell that the numbers here are related to the numbers there unless you read the code quite carefully

someone updating the tests later could change the numbers here and not realise they needed to change the numbers there

i'd recommend defining a constant for this `WeeklyOccupancy` instance in the test class

SeriousGuy888 commented on src/test/java/use_case/autogen/AutogenInteractorTest.java:
```diff
@@ -0,0 +1,364 @@
+
+        interactor.execute(inputData);
+
+        System.out.println("\n*** ALWAYS PRINTING RESULT ***");
```
im confused: why does this say `ALWAYS PRINTING RESULT` instead of just `GENERATION RESULT` or the name of the test that was just run or something?

SeriousGuy888 reviewed (changes requested):
$\huge{\tiny\text{H}\small\text{u}\normalsize\text{z}\large\text{z}\LARGE\text{a}\huge\text{h}\huge!\huge\large!\huge\LARGE!}$ Abish has mostly implemented the timetable autogeneration feature$`\small!\normalsize!\large!\LARGE!\huge\large!`$

i may have written a lot of comments in the code
but they're mostly style concerns or typo fixes

the main thing that actually _matters_ is
im a bit iffy still on the `CourseVariableFactory` class
(see comments there)

most things are good though
i don't think there's anything major to change

Abish-27 committed:
Got rid of CourseVariableFactory.java and replaced it with a applyLock() function in AutogenInteractor

Abish-27 committed:
Update src/main/java/entity/constraints/TimeConflictConstraint.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Update src/main/java/interface_adapter/autogen/AutogenPresenter.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Update src/main/java/interface_adapter/autogen/AutogenPresenter.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Update src/test/java/use_case/autogen/AutogenInteractorTest.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Update src/main/java/use_case/autogen/AutogenDataAccessInterface.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Update src/main/java/use_case/autogen/AutogenInputData.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Victoriaecai reviewed (dismissed):
Looks good! I agree with Bill's suggestions and just had a couple comments about style but otherwise is looking great

Abish-27 committed:
Update src/main/java/use_case/autogen/AutogenInteractor.java

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Changed output data of Set of sections to Timetable object for autogen.

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Changed AutogenDataAccessInterface to take in sections rather than input data.

Abish-27 committed:
Merge remote-tracking branch 'origin/refactor-WeeklyOccupancy-to-use-integer-lists-of-milliseconds-instead' into autogen

# Conflicts:
#	pom.xml

Abish-27 committed:
Adjusted AutogenInteractorTest for revised presenter/OutputData

Abish-27 committed:
No more magic numbers in AutogenInteractorTest

Abish-27 committed:
No more one-liners in code.

Abish-27 committed:
Refactored PotentialTimetable to entity package.

Abish-27 commented:
Finished implementing all changes requested by Bill and Victoria. PR once again ready for review.

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Updated AutogenViewModel to correct format

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Now Autogen also checks for semester clash

SeriousGuy888 commented on pom.xml:
```diff
@@ -13,6 +13,14 @@
+            <version>5.8.1</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
duplicate dependencies

```suggestion
```

Abish-27 committed:
Updated tests to account for semester. Add new semester test as well.

Abish-27 committed:
Update pom.xml

Co-authored-by: SeriousGuy888 <38299346+SeriousGuy888@users.noreply.github.com>

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

SeriousGuy888 reviewed (approved):
new changes look good i think

SeriousGuy888 commented:
the `WeeklyOccupancy` problem might clash with #65 though

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
minor style updates

Abish-27 committed:
Merge branch 'timetable-view' into autogen

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Merge branch 'connect-search-panel' into autogen

Abish-27 committed:
Added AutogenDAO

Abish-27 committed:
Added new tests that check if DAO works (still more to add)

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Merge remote-tracking branch 'refs/remotes/origin/usecase-save-and-load-workbook-with-ui' into autogen

Abish-27 committed:
Connected AutogenInteractor To UI (finally)

Abish-27 committed:
Changed autogen tests

Abish-27 committed:
Merge branch 'main' into autogen

SeriousGuy888 committed:
fixed broken SaveWorkbookInteractorTest

it was broken because SaveWorkbookInteractor no longer gives that
exact error message.
somehow abish managed to merge in part of my modified save usecase
but not the updated tests?

Abish-27 committed:
Merge branch 'usecase-save-and-load-workbook-with-ui' into autogen

# Conflicts:
#	src/main/java/app/AppBuilder.java

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Something

Abish-27 committed:
added locked section options to Timetable entity

Abish-27 committed:
Added LockSection Usecase files

Abish-27 committed:
Added LockSection Usecase files part 2

Abish-27 committed:
Created LockSection Controllers/Presenters and connected to UI

Abish-27 committed:
Updated Lock Sections Table

Abish-27 committed:
Updated AutogenController

Abish-27 committed:
Finished LockSection Table UI

Abish-27 committed:
Changed AppBuilder DAO initialisation

Abish-27 committed:
x

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Merge branch 'main' into usecase-locksections-revised

Abish-27 committed:
Merge branch 'refs/heads/usecase-locksections-revised' into autogen

# Conflicts:
#	.idea/workspace.xml

Victoriaecai committed:
Merge pull request #74 from SeriousGuy888/fix-unrelated-broken-test-on-autogen

fixed broken SaveWorkbookInteractorTest in autogen branch

Abish-27 committed:
Finished Implementing AutogenController

Abish-27 committed:
Merge remote-tracking branch 'origin/autogen' into autogen

Abish-27 committed:
Minor style changes

Abish-27 committed:
Deleted unnecessary AutogenViewModel

Abish-27 committed:
Merge branch 'main' into autogen

Abish-27 committed:
Merge remote-tracking branch 'origin/autogen' into autogen

# Conflicts:
#	.idea/workspace.xml

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'refs/heads/main' into autogen

# Conflicts:
#	.idea/workspace.xml
#	src/main/java/app/AppBuilder.java

Abish-27 committed:
OMGGGG ITS WORKING

CarolineLyu reviewed (approved):
good work this was heavy lift

Abish-27 committed:
Merge remote-tracking branch 'origin/main'

Abish-27 committed:
Merge branch 'main' into autogen
